packages/malloy/src/malloy.ts

```diff
@@ -3882,49 +3904,152 @@ export class CSVWriter extends DataWriter {
-      return cell.value.toISOString();
-    } else if (cell.isString()) {
-      return cell.value;
+    } else if (value instanceof Date) {
+      return value.toISOString();
+    } else if (typeof value === 'boolean' || typeof value === 'number') {
+      return JSON.stringify(value);
     } else {
-      return `${cell.value}`;
+      return `${value}`;
     }
   }
 
-  async process(data: AsyncIterableIterator<DataRecord>): Promise<void> {
-    let fields;
-    for await (const row of data) {
-      if (fields === undefined) {
-        fields = row.field.allFields;
-        if (this.includeHeader) {
-          for (let fieldIndex = 0; fieldIndex < fields.length; fieldIndex++) {
-            const field = fields[fieldIndex];
-            this.stream.write(this.escape(field.name));
-            if (fieldIndex !== fields.length - 1) {
-              this.stream.write(this.columnSeparator);
-            }
-          }
-          this.stream.write(this.rowSeparator);
+  // Extra weight to be added becase of nested tables inside the cells.
+  private getColWeight(jsonVal: QueryDataRow | QueryData) {
+    let firstVal = jsonVal;
+    if (Array.isArray(jsonVal)) {
+      firstVal = jsonVal[0];
+    }
+    let numKeys = 0;
+    for (const key in firstVal) {
+      numKeys = numKeys + 1;
+      const val = firstVal[key];
+      if (Array.isArray(val)) {
+        const weight = this.getColWeight(val) - 1;
+        numKeys = numKeys + weight;
+      }
+    }
+    return numKeys;
+  }
+
+  // Get header row along with extra empty spaces for nested children.
+  private getHeaderRow(row: QueryDataRow): CellMatrix {
+    let csv = '';
+    let width = 0;
+    for (const key in row) {
+      csv = csv + this.escape(key) + this.columnSeparator;
+      const val = row[key];
+      width++;
+      if (Array.isArray(val)) {
+        const numKeys = this.getColWeight(val) - 1;
+        width = width + numKeys;
+        for (let i = 0; i < numKeys; i++) {
+          csv = csv + this.columnSeparator;
         }
       }
-      for (let fieldIndex = 0; fieldIndex < fields.length; fieldIndex++) {
-        const field = fields[fieldIndex];
-        this.stream.write(this.escape(this.stringify(row.cell(field))));
-        if (fieldIndex !== fields.length - 1) {
-          this.stream.write(this.columnSeparator);
+    }
+    return {rows: [csv], length: 1, width: width};
+  }
+
+  // Merge the child matrices i.e. merge the columns into one bigger matrix i.e. CSV.
+  private mergeMatrices(matrices: CellMatrix[]): CellMatrix {
+    const maxLength = Math.max(...matrices.map(matrix => matrix.length));
+    const matrixWidth = matrices.reduce((sum, matrix) => sum + matrix.width, 0);
+    const csvMatrix: string[] = [];
+    for (let i = 0; i < maxLength; i++) {
+      let csvRow = '';
+      for (const matrix of matrices) {
+        if (i < matrix.length) {
+          csvRow = csvRow + matrix.rows[i];
+          if (!csvRow.endsWith(this.columnSeparator)) {
```

Comment:
Is the goal here to not have two separators in a row? That seems like it could get confused by empty cells. In general, the more efficient and predictable way to separate a series of cells is to create an array and then do `.join()`. That prevents trailing separators like the ones that are failing tests now.

If you can compute the length of the array in advance you can save some memory allocations, but even if you can't `.push()` will overall be more performant and memory efficient.

Reply:
Done